cranelift/codegen/src/isa/x86/abi.rs

```diff
@@ -834,19 +850,52 @@ fn insert_common_prologue(
 /// compared to the stack pointer, but currently it serves enough functionality
 /// to get this implemented in `wasmtime` itself. This'll likely get expanded a
 /// bit over time!
-fn interpret_gv(pos: &mut EncCursor, gv: ir::GlobalValue, scratch: ir::ValueLoc) -> ir::Value {
+fn interpret_gv(
+    pos: &mut EncCursor,
+    gv: ir::GlobalValue,
+    sp: Option<ir::Value>,
+    scratch: ir::ValueLoc,
+) -> ir::Value {
     match pos.func.global_values[gv] {
-        ir::GlobalValueData::VMContext => pos
-            .func
-            .special_param(ir::ArgumentPurpose::VMContext)
-            .expect("no vmcontext parameter found"),
+        ir::GlobalValueData::VMContext => {
+            let vmctx_index = pos
+                .func
+                .signature
+                .special_param_index(ir::ArgumentPurpose::VMContext)
+                .expect("no vmcontext parameter found");
+            match pos.func.signature.params[vmctx_index] {
+                AbiParam {
+                    location: ArgumentLoc::Reg(_),
+                    ..
+                } => {
+                    let entry = pos.func.layout.entry_block().unwrap();
+                    pos.func.dfg.block_params(entry)[vmctx_index]
+                }
+                AbiParam {
+                    location: ArgumentLoc::Stack(offset),
+                    value_type,
+                    ..
+                } => {
+                    let offset = offset + i32::from(pos.isa.pointer_bytes());
+                    let ret =
+                        pos.ins()
+                            .load(value_type, ir::MemFlags::trusted(), sp.unwrap(), offset);
+                    pos.func.locations[ret] = scratch;
+                    return ret;
```

Comment:
You should use `pos.func.stack_slots.make_incoming_arg(types::I32, offset)` and then load from the resulting stack slot.

Reply:
Does that actually work in the prologue itself?

Reply:
Ish. Stack slots are handled correctly in x86 after rsp's final adjustment, even though there may be a little more prologue afterward (for example, x86_64 XMM preservation [relies on this](https://github.com/bytecodealliance/wasmtime/blob/d2ec33c6caec7eda7afcf9f025c88b413b7ff0f1/cranelift/codegen/src/isa/x86/abi.rs#L799)). It looks like this is too early for an explicit `stack_load` here.

Reply:
That was my understanding too, which is why I spent several hours figuring out a solution that does not involve a `stack_load`.

Reply:
Maybe use `fill`? That one takes the SSA var stored on stack and outputs it ib a register.

Reply:
@bjorn3 `fill` has the exact same problem as `stack_load`, no? It also uses stack slots, which is what I am avoiding in this code.

Reply:
@iximeow Could you advise on any workable way to implement this operation? I'll be happy to implement whatever the upstream likes here. I think I could use a `stack_load` with an ad-hoc fixup if that's considered OK, though that looks pretty gross to me personally.

Reply:
I think this is the workable approach, since stack slot machinery isn't usable this early in the function. I agree that `stack_load` with some later fixup would not be better.